crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -264,13 +265,12 @@ async fn app_validation_workflow_inner(
             t,
             a,
             r,
-            total
+            ops_validated
         );
     }
     jh.await?;
-    tracing::debug!("accepted {} ops", total);
-    Ok(if saturated {
-        WorkComplete::Incomplete
+    Ok(if saturated || ops_validated < num_ops_to_validate {
```

Comment:
This seems a bit risky to me. This is the count of accepted ops. Yes that works well when we only get accepted or awaiting outcomes but what about rejected? If we rejected ops we don't count them towards the total. Poor variable naming isn't helping here! `total` really ought to be a total or be renamed to `accepted`.

I think it would be safer to use a check for 'awaiting deps' count > 0. But even then, we don't know when those will arrive or if they ever will. We're basically introducing a loop that will run every 2.5s in every Holochain instance that has a permanently missing dep. I'm not aware there's currently any 'discard' logic if an op has too many validation attempts.

I actually wonder if it'd be better to make this a deliberate loop that checks for work and does validation if there is work to do and stops quickly if not.

Reply:
I agree to your concern and realized too that the counters are not reliable. I do not plan to merge this PR any longer, as it introduces repercussions we haven't tested yet, and would much prefer to improve our workflow logic.

Reply:
@jost-s it seems you've revised the PR now, but this concern still remains.

Maybe you could also include a `HashMap<DhtOpHash, usize>` in the Incomplete variant to track the ops awaiting validation, along with the number of validation attempts (maybe let validation be attempted for 1 minute, or until?) before ceasing to retrigger? The hashmap would have to be fed back into the workflow. This would still result in validation attempts for the awaiting ops the next time any other op comes in, but at least it wouldn't be an endless loop even when holochain is idle.

Reply:
This sounds like a good suggestion. Either this or including the rejected ops in the count would convince me of this logic

Reply:
The rejected ops are counted in now. Sorry @maackle and @ThetaSinner, I hadn't put your comments into code before asking for review again.

Our interim decision was to have this endless loop without condition, but I'll see how much refactoring Michael's suggestion will be.

Reply:
@maackle I've looked at implementing your suggestion, but I'd have to refactor the workflow trigger to accept the hash map to be passed in AFAICS. Another obstacle is that the ops are not validated individually but in chunks and that would have to be refactored too to have the 1 minute waiting period.

All in all I don't feel confident to refactor that without testing it and don't think it's worth putting in the effort if right after that I start working on the actual refactor of the app validation.

Reply:
This looks safe to me now, I'm happy :)